Precommit addition to the repo

ryandanehy opened the pull request:
https://pre-commit.ci/

[clang format style](https://releases.llvm.org/12.0.0/tools/clang/docs/ClangFormatStyleOptions.html
)
[cmake style formatting [docs]](https://pypi.org/project/cmake-format/)

Fixes #5 
Fixes #54

pelesh commented:
@ryandanehy, can we stage this PR? First stage would be to set up configurations. Once that is merged, we can run formatting on the code and have a separate PR to merge formatted code. Is that doable?

ryandanehy commented:
> @ryandanehy, can we stage this PR? First stage would be to set up configurations. Once that is merged, we can run formatting on the code and have a separate PR to merge formatted code. Is that doable?

Once we merge this any push to a PR that rebased on Develop will automatically format the code. I can change the CI to only run when manually triggered if that is preferred?

Essentially options are manual trigger or automatically with every push to a PR.

pelesh commented:
> @pelesh I think this is good to go with the merge. Then I will open a new pr with all the precommit suggested code changes. The precommit ci fails currently because it thinks the code is not format correctly so it is doing it jobs correctly.

Let's review. I propose we merge it after #83 and #86 are merged.

cameronrutherford commented on .github/workflows/pre_commit.yaml:
```diff
@@ -0,0 +1,29 @@
+        ref: ${{ github.event.pull_request.head.ref }}
+    - uses: actions/setup-python@v3
+    - uses: pre-commit/action@v3.0.0
+    # - uses: EndBug/add-and-commit@v9.1.3
+    #   # Only need to try and commit if the action failed
+    #   if: failure()
+    #   with:
+    #     fetch: false
+    #     committer_name: GitHub Actions
+    #     committer_email: actions@github.com
+    #     message: Apply pre-commmit fixes 
```
I think we want pre-commit to automatically apply fixes, so lets comment this back in

cameronrutherford commented on .github/workflows/pre_commit.yaml:
```diff
@@ -0,0 +1,29 @@
+
+# Won't run on develop/main directly
+on:
+ workflow_dispatch: 
+    inputs:
+      branch:
+        description: 'The branch to build'
+        required: true
```
```suggestion

```
If we just run on `pull_request`, doesn't that generate the desired behavior?

ryandanehy replied:
I was trying to set it to manual so that people could start the workflow as needed. This is in reference to https://github.com/ORNL/ReSolve/pull/82#issuecomment-1832300152.

cameronrutherford replied:
Well I suggest maybe deferring the implementation of that new feature into a different PR since it doesn't seem to work here, and is unclear at the moment.

Maybe instead you could also code a GitHub bot that can run the fixes for us automatically if you really want to over-engineer this...

cameronrutherford replied:
I still don't get this. Why not just on `pull_request`... Your code seems like it's non-functional as I have never seen manual pipelines before

cameronrutherford commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,28 @@
+ci:
+    autofix_commit_msg: |
+        [pre-commit.ci] auto fixes from pre-commit.com hooks
+
+        for more information, see https://pre-commit.ci
+    autofix_prs: true
+    autoupdate_branch: ''
+    autoupdate_commit_msg: '[pre-commit.ci] pre-commit autoupdate'
+    autoupdate_schedule: weekly
+    skip: []
+    submodules: false
```
I'm not familiar with this configuration section. Shouldn't we just use what's in the GitHub action to autofix, and then customize the commit message there?

cameronrutherford replied:
I still don't think this configuration section is necessary at all. The commit message is controlled by the GitHub action, and this does nothing.

cameronrutherford reviewed (changes requested):
A few changes, but as well we should verify this auto-fixes changes before merging. Your pipelines are failing, and they should be abl to fix themselves.

cameronrutherford commented:
> @cameronrutherford  I commented out the auto fixes per slaven request for this PR. See https://github.com/ORNL/ReSolve/pull/82#issuecomment-1831224589 
> 
> 
> 
> The pipelines fail because it wants to autofix but I am not letting it. 

I guess we merge this with failing pipelines then :(

pelesh commented:
> > @cameronrutherford  I commented out the auto fixes per slaven request for this PR. See [#82 (comment)](https://github.com/ORNL/ReSolve/pull/82#issuecomment-1831224589)
> > The pipelines fail because it wants to autofix but I am not letting it.
> 
> I guess we merge this with failing pipelines then :(

The other option is to enable auto checks and we merge this PR last thing before the release. Once we turn on style auto-correction all rebasing and merging will turn into a nightmare.

cameronrutherford commented:
> > > @cameronrutherford  I commented out the auto fixes per slaven request for this PR. See [#82 (comment)](https://github.com/ORNL/ReSolve/pull/82#issuecomment-1831224589)
> 
> > > The pipelines fail because it wants to autofix but I am not letting it.
> 
> > 
> 
> > I guess we merge this with failing pipelines then :(
> 
> 
> 
> The other option is to enable auto checks and we merge this PR last thing before the release. Once we turn on style auto-correction all rebasing and merging will turn into a nightmare.

Since you can now rebase automatically with buttons, it's not so bad. I do agree though that with PRs in motion it would be easier to merge this after those.

ryandanehy commented:
@pelesh I rebased, added some instructions in the main readme.md, and fixed pre-commit so that it does not sort includes. Let me your thoughts so we can get this merged.

cameronrutherford commented on README.md:
````diff
@@ -53,7 +53,27 @@ target_link_libraries(my_app PRIVATE ReSolve::ReSolve)
 
+Precommit is auto applied through CI to every pull request and every commit to the said pull request. Precommit checks for valid yaml files, end of file spaces, Syntax of C codeand etc. To see specifically what formating precommit looks for please check out our .clang-format and .cmake-format filed as well as our .pre-commit-config.yaml which show the specific precommit hooks being applied. 
+
+### Running Pre-commit locally
+To run pre-commit locally you first must install pre-commit. It can be installed via
+```shell
+pip install pre-commit
+```
+
````
How do I install pip? What about Python? What if I am on a linux vs mac?

cameronrutherford commented on README.md:
````diff
@@ -53,7 +53,27 @@ target_link_libraries(my_app PRIVATE ReSolve::ReSolve)
+
+To install the hooks and have pre-commit run automatically before every commit run
+```shell
+pre-commit install
````
```suggestion
pre-commit install --install-hooks
```

cameronrutherford commented on README.md:
````diff
@@ -53,7 +53,27 @@ target_link_libraries(my_app PRIVATE ReSolve::ReSolve)
+```shell
+pre-commit run --all-files
+```
+To learn more about pre-commit usuage check out https://pre-commit.com/#usage.
````
Maybe add some more information here about:
- Rebasing against auto-fixes in PRs
- Understanding git-flow in this context
  - What do I do when pre-commit fails?
  - Why does all this new stuff appear when I run `git commit`?

cameronrutherford reviewed (changes requested):
Getting there

pelesh commented:
I believe the failed test on Ascent is due to issue fixed in #109. Perhaps a good idea would be rebasing to `develop` to make sure this test passes?

ryandanehy commented:
> I believe the failed test on Ascent is due to issue fixed in #109. Perhaps a good idea would be rebasing to `develop` to make sure this test passes?

Just rebased.

cameronrutherford commented:
pre-commit tests are still failing and no auto-fix is being applied
